File: scala/173.scala
import scala.collection.mutable.Stack
class BSTIterator(_root: TreeNode) {

    val s = {
        var ptr = _root
        val stack = Stack[TreeNode]()
        while (ptr != null) {
            stack.push(ptr)
            ptr = ptr.left
        }
        stack
    }

    def next(): Int = {
        val cur = s.pop()
        var ptr = cur.right
        while (ptr != null) {
            s.push(ptr)
            ptr = ptr.left
        }
        cur.value
    }

    def hasNext(): Boolean = {
        !s.isEmpty
    }

}

File: leetcodeJava/173.java
package leetcodeJava;

class BSTIterator {
    
    LinkedList<TreeNode> stack;

    public BSTIterator(TreeNode root) {
        stack = new LinkedList<TreeNode>();
        var ptr = root;
        while(ptr != null) {
            stack.add(ptr);
            ptr = ptr.left;
        }
    }
    
    public int next() {
        var n = stack.removeLast();
        if (n.right != null) {
            var ptr = n.right;
            while (ptr != null) {
                stack.add(ptr);
                ptr = ptr.left;
            }
        }
        return n.val;
    }
    
    public boolean hasNext() {
        return !stack.isEmpty();
    }
}
